### PR TITLE
Impl PartialOrd<&B> for &mut A

### PR DESCRIPTION
**Pull request by Duddino**

### CODE REVIEW

**Comment by ollie27** [comment]
How about adding `PartialOrd<&mut B> for &A` as well to match `PartialEq`?

**Comment by Duddino** [comment]
But you can already compare &T with &mut T

**Comment by ollie27** [comment]
> 
> 
> But you can already compare &T with &mut T

It depends how you call it. `PartialOrd::lt(&&0u8, &&mut 0u8)` doesn't work even though `PartialEq::eq(&&0u8, &&mut 0u8)` does.

Implementing `PartialOrd` symmetrically is actually [required](https://doc.rust-lang.org/nightly/std/cmp/trait.PartialOrd.html):

>Note that these requirements mean that the trait itself must be implemented symmetrically and transitively: if `T: PartialOrd<U>` and `U: PartialOrd<V>` then `U: PartialOrd<T>` and `T: PartialOrd<V>`.

**Comment by Duddino** [comment]
@ollie27 Thank you, I should have fixed it

**Comment by Duddino** [comment]
I get an error when trying to impl PartialOrd<&mut B> for &B:
```
error[E0277]: can't compare `Val` with `&Val`
   --> /home/duddino/.cargo/registry/src/github.com-1ecc6299db9ec823/datafrog-2.0.1/src/treefrog.rs:379:50
    |
379 |                 slice = gallop(slice, |kv| &kv.1 < v);
    |                                                  ^ no implementation for `Val < &Val` and `Val > &Val`
    |
    = help: the trait `std::cmp::PartialOrd<&Val>` is not implemented for `Val`
    = help: consider adding a `where Val: std::cmp::PartialOrd<&Val>` bound
    = note: required because of the requirements on the impl of `std::cmp::PartialOrd<&&Val>` for `&Val`

    Checking rustc_lexer v0.1.0 (/home/duddino/Desktop/rust/src/librustc_lexer)
    Checking chalk-macros v0.1.0
error[E0277]: can't compare `Val` with `&Val`
   --> /home/duddino/.cargo/registry/src/github.com-1ecc6299db9ec823/datafrog-2.0.1/src/treefrog.rs:464:54
    |
464 |                     slice = gallop(slice, |kv| &kv.1 < v);
    |                                                      ^ no implementation for `Val < &Val` and `Val > &Val`
    |
    = help: the trait `std::cmp::PartialOrd<&Val>` is not implemented for `Val`
    = help: consider adding a `where Val: std::cmp::PartialOrd<&Val>` bound
    = note: required because of the requirements on the impl of `std::cmp::PartialOrd<&&Val>` for `&Val`

    Checking libc v0.2.62
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `datafrog`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by ollie27** [comment]
> I get an error when trying to impl PartialOrd<&mut B> for &A:

I'm not completely sure what's going on there. I think there is something weird with binary operators. For some reason `&0u8 < &&0u8` currently compiles but `&0u8 == &&0u8` doesn't with a similar error message to what you're seeing.

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -1208,6 +1208,36 @@ mod impls {
...
+        fn le(&self, other: &&B) -> bool { PartialOrd::le(*self, *other) }
+        #[inline]
+        fn ge(&self, other: &&B) -> bool { PartialOrd::ge(*self, *other) }
+        #[inline]
+        fn gt(&self, other: &&B) -> bool { PartialOrd::gt(*self, *other) }
+    }
+    #[stable(feature = "partial_ord_mut_ref", since = "1.40.0")]
+    impl<A: ?Sized, B: ?Sized> PartialOrd<&mut B> for &A where A: PartialOrd<B> {
+        #[inline]
+        fn partial_cmp(&self, other: &&B) -> Option<Ordering> {
```
Comment (ollie27) [inline src/libcore/cmp.rs]:
```suggestion
        fn partial_cmp(&self, other: &&mut B) -> Option<Ordering> {
```

and the same for the methods below.

**Comment by Duddino** [comment]
I should have been more specific, sorry! I already figured how to correctly implement it, but it gives me another on my local ./x.py check

**Comment by Duddino** [comment]
@ollie27 so probably PartialEq<&B> is preventing &a == &&b to compile

**Comment by ollie27** [comment]
> so probably impl PartialEq<&B> for &mut A is preventing &a == &&b to compile (?)

Well `impl PartialEq<&mut B> for &A` but yeah that's my guess.

**Comment by Alexendoo** [comment]
Ping from triage, @Duddino what's the status of this PR, still having trouble with https://github.com/rust-lang/rust/pull/65619#issuecomment-544278250? Thanks

**Comment by joelpalmer** [comment]
Ping from Triage: Closing due to inactivity. @Duddino, please re-open with updates. Thank you for the PR.

**Pull request closed without merging**